src/poetry/utils/authenticator.py

```diff
@@ -298,13 +325,37 @@ def _get_credentials_for_netloc_from_keyring(
 
         return None
 
-    def _get_certs_for_netloc_from_config(self, netloc: str) -> dict[str, Path | None]:
-        certs: dict[str, Path | None] = {"cert": None, "verify": None}
+    @functools.lru_cache(maxsize=None)
+    def _get_repository_config_for_url(
+        self, url: str
+    ) -> AuthenticatorRepositoryConfig | None:
+        parsed_url = urllib.parse.urlsplit(url)
+        candidates_netloc_only = []
+        candidates_path_match = []
+
+        for repository in self.configured_repositories.values():
+            if repository.netloc == parsed_url.netloc:
+                if parsed_url.path.startswith(repository.path):
+                    candidates_path_match.append(repository)
+                    continue
+                candidates_netloc_only.append(repository)
+
+        if candidates_path_match:
+            candidates_path_match.sort(key=lambda c: len(c.path), reverse=True)
+            return candidates_path_match[0]
+        elif candidates_netloc_only:
+            if len(candidates_netloc_only) > 1:
+                logger.debug(
+                    "Multiple source configurations found for %s - %s",
+                    parsed_url.netloc,
+                    ", ".join(map(lambda c: c.name, candidates_netloc_only)),
+                )
```

Comment:
```suggestion
            if len(candidates_netloc_only) > 1:
                from os.path import commonprefix
                candidates_netloc_only.sort(key=lambda c: len(commonprefix([parsed_url.path, c.path])), reverse=True)
                logger.debug(
                    "Multiple source configurations found for %s - %s",
                    parsed_url.netloc,
                    ", ".join(map(lambda c: c.name, candidates_netloc_only)),
                )
```

Apparently Azure DevOps Artifact Feeds seem to prefer using the guid instead of the Feed's display name.
```
https://pkgs.dev.azure.com/<ORG NAME>/_packaging/<FEED DISPLAY NAME>/pypi/simple/<PACKAGE>
```
ends up downloading
```
https://pkgs.dev.azure.com/<ORG NAME>/_packaging/<GUID>/pypi/simple/<PACKAGE>/<VERSION>/<WHEEL>#sha256=<HASH>
```
resulting in the str.startswith() check to return False.

My suggestion sort on the longest common prefix for the `path` component of the url as a best effort.

(Azure DevOps Personal Access Tokens are scroped to an organization, not to individual projects in an organization, so this works)